crates/polars-parquet/src/parquet/encoding/hybrid_rle/encoder.rs

```diff
@@ -3,98 +3,232 @@ use std::io::Write;
     writer: &mut W,
     iterator: I,
+    num_bits: u32,
 ) -> std::io::Result<()> {
-    // the length of the iterator.
-    let length = iterator.size_hint().1.unwrap();
-
-    // write the length + indicator
-    let mut header = ceil8(length) as u64;
-    header <<= 1;
-    header |= 1; // it is bitpacked => first bit is set
-    let mut container = [0; 10];
-    let used = uleb128::encode(header, &mut container);
-
-    writer.write_all(&container[..used])?;
-
-    // encode the iterator
-    bitpacked_encode(writer, iterator)
+    let mut consecutive_repeats: usize = 0;
+    let mut previous_val = T::default();
+    let mut buffered_bits = [previous_val; MAX_VALUES_PER_LITERAL_RUN];
+    let mut buffer_idx = 0;
+    let mut literal_run_idx = 0;
+    for val in iterator {
+        if val == previous_val {
+            consecutive_repeats += 1;
+            if consecutive_repeats >= 8 {
+                // Run is long enough to RLE, no need to buffer values
+                if consecutive_repeats > 8 {
+                    continue;
+                } else {
+                    // When we encounter a run long enough to potentially RLE,
+                    // we must first ensure that the buffered literal run has
+                    // a multiple of 8 values for bit-packing. If not, we pad
+                    // up by taking some of the consecutive repeats
+                    let literal_padding = (8 - (literal_run_idx % 8)) % 8;
+                    consecutive_repeats -= literal_padding;
+                    literal_run_idx += literal_padding;
+                }
+            }
+            // Too short to RLE, continue to buffer values
+        } else if consecutive_repeats > 8 {
+            // Value changed so start a new run but the current run is long
+            // enough to RLE. First, bit-pack any buffered literal run. Then,
+            // RLE current run and reset consecutive repeat counter and buffer.
+            if literal_run_idx > 0 {
+                debug_assert!(literal_run_idx % 8 == 0);
+                T::bitpacked_encode(
+                    writer,
+                    buffered_bits.iter().take(literal_run_idx).copied(),
+                    num_bits as usize,
+                )?;
+                literal_run_idx = 0;
+            }
+            T::run_length_encode(writer, consecutive_repeats, previous_val, num_bits)?;
+            consecutive_repeats = 1;
+            buffer_idx = 0;
+        } else {
+            // Value changed so start a new run but the current run is not long
+            // enough to RLE. Consolidate all consecutive repeats into buffered
+            // literal run.
+            literal_run_idx = buffer_idx;
+            consecutive_repeats = 1;
+        }
+        // If buffer is full, bit-pack as literal run and reset
+        if buffer_idx == MAX_VALUES_PER_LITERAL_RUN {
+            T::bitpacked_encode(writer, buffered_bits.iter().copied(), num_bits as usize)?;
```

Comment:
The issue with my original PR was here. Instead of `buffered_bits.iter().copied()`, I had `buffered_bits.iter().copied().take(literal_run_idx)`.

When the buffer filled up in the middle of a consecutive run, `literal_run_idx != MAX_VALUES_PER_LITERAL_RUN` so data was lost. The solution was to remove the `take` clause and always flush the entire buffer when full.

I added a `debug_assert` to check that proper number repeats are consolidated into the buffered literal run.